Reinforce the "save_to" method to checkout the seekability when download files

athakur-reef opened the pull request.

mjurbanski-reef commented on b2sdk/transfer/inbound/downloaded_file.py:
```diff
@@ -203,3 +203,24 @@ def save_to(self, path_, mode='wb+', allow_seeking=True):
+        :param allow_seeking: if False, download strategies that rely on seeking to write data
+                              (parallel strategies) will be discarded.
+        """
+        try:
+            self.save_to(path_, mode, allow_seeking)
+        except io.UnsupportedOperation:
+            if mode != 'wb+':
+                raise
+            self.save_to(path_, 'wb', allow_seeking)
```
pretty sure this is not going to work depending on the download strategy, if not for all of them with sufficiently large file

also why new method if this is `safe`?

athakur-reef replied:
- Of course, I'm not sure this is working for all cases. So I tried to test this code for several cases such as `Parallel` .
And also tested on `multi-thread` mode and confirmed everything is okay. Because of this, I changed the `allow_seeking`.
- And regardig the why I created the `safe` mode, beceause this issues is arising for non-seekable files.
I think try to download files to non-seekable file such as `/dev/stdout` is one of rarly cases. and it also affect to the performance and I think we can use original `save_to` function for eager running mode.

mjurbanski-reef replied:
current implementation may cause issues:
* we have new method, which IMO doesn't quite make sense, if we can support all in one - this will just confuse users and give more work for maintainers
* if UnsupportedOperation (Which is quite vague) is raised but anything other than `open` it will cause very nasty and hard to detect bugs, since part of the download stream will be consumed by first try

better to figure out why we need `wb+` - I'm not sure if we need it at all, or will `+` suffice.
Maybe if allow_seeking is True requires `wb+`, but otherwise, we should do without it IMO.

athakur-reef commented:
- Implemented test
- Disabled `allow_seeking` for non-seekable output
- Updated CHANGELOG

athakur-reef commented:
- Implemented test
- Disabled `allow_seeking` for non-seekable output
- Updated CHANGELOG
- Remove `safe_save_to` method again.
- Reinforce the `save_to` method to check the readability and ability of parallel strategy

mjurbanski-reef commented on b2sdk/transfer/inbound/downloaded_file.py:
```diff
@@ -50,12 +53,13 @@ class MtimeUpdatedFile(io.IOBase):
     """
 
-    def __init__(self, path_, mod_time_millis: int, mode='wb+', buffering=None):
+    def __init__(self, path_, mod_time_millis: int, preferred_mode='wb+', buffering=None):
```
please keep the old name

I don't see a good reason to introduce a API breaking change here. Just make default `None` for ppl that did not care about the exact mode set.

mjurbanski-reef commented on b2sdk/transfer/inbound/downloaded_file.py:
```diff
@@ -165,6 +174,9 @@ def save(self, file, allow_seeking=True):
         :param allow_seeking: if False, download strategies that rely on seeking to write data
                               (parallel strategies) will be discarded.
         """
+        if file.mode == 'wb':
+            allow_seeking = False
```
I don't think silently overwritting argument is a good idea, especially since whole thing can be easily overriden by us on CLI call for our purposes.

So it should other way around, i.e.

If `mode` wasn't set explicitly by the user, then: if `allow_seeking` is explicitly set by someone to `False`, we should use the default `wb`, and otherwise, use `wb+`.

Lets not forget our goal, which is to make download to `/dev/stdout` possible through CLI. And that is all.

In fact that, after looking at it more, it seems to me that this goal requires no changes to SDK, since we could very well do detection of "streamable" output file in CLI, and just call `.save_to(..., mode="wb", allow_seeking=False)`.

It seems all SDK changes can be forgone and this PR closed.

ppolewicz replied:
I disagree. I think the sdk, when it sees a named pipe or a standard output (a device?), should switch just to a non-seekable output mode automatically... unless the user sets `allow_seeking` explicitly to `True` or `False`, in which case it's their fault if they configure it in a way that then doesn't work.

mjurbanski-reef commented on b2sdk/transfer/inbound/downloaded_file.py:
```diff
@@ -157,14 +178,23 @@ def _validate_download(self, bytes_read, actual_sha1):
                 raise TruncatedOutput(bytes_read, desired_length)
 
-    def save(self, file, allow_seeking=True):
+    def save(self, file, allow_seeking: bool = None):
```
please fix type annotation; should be Optiona[bool]

mjurbanski-reef commented on b2sdk/transfer/inbound/downloaded_file.py:
```diff
@@ -187,7 +217,7 @@ def save(self, file, allow_seeking=True):
         self._validate_download(bytes_read, actual_sha1)
 
-    def save_to(self, path_, mode='wb+', allow_seeking=True):
+    def save_to(self, path_, mode: Literal['wb', 'wb+'] = None, allow_seeking: bool = None):
```
missing Optional[]

mjurbanski-reef commented on b2sdk/transfer/inbound/downloaded_file.py:
```diff
@@ -157,14 +178,23 @@ def _validate_download(self, bytes_read, actual_sha1):
         :param allow_seeking: if False, download strategies that rely on seeking to write data
                               (parallel strategies) will be discarded.
         """
+        if allow_seeking and file.seekable():
```
:bug:

mjurbanski-reef commented on b2sdk/transfer/inbound/downloaded_file.py:
```diff
@@ -157,14 +178,22 @@ def _validate_download(self, bytes_read, actual_sha1):
+        if allow_seeking and not file.seekable():
+            logger.warning('You allowed seeking for non-seekable file')
+        elif allow_seeking is None:
+            if file.mode == 'wb':
+                allow_seeking = False
+            else:
+                allow_seeking = True
```
```suggestion
            allow_seeking = file.seekable()
```

mjurbanski-reef commented on CHANGELOG.md:
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
+### Fixed
+
+* Fixed to check the readability and parallel strategy for download
```
```suggestion
* Support stdout and non-seekable files as download output target
```

ppolewicz commented on b2sdk/transfer/inbound/downloaded_file.py:
```diff
@@ -93,12 +108,21 @@ def __enter__(self):
+                buffering=self.buffering,
+            )
+        except io.UnsupportedOperation:
+            if self.mode is not None:
+                raise
```
if someone asked to open the file in a specific mode and that can't be done, error out. You shouldn't be trying to repair it as it's unexpected behavior, someone may be running this on top of a fuse which does something funny on open(). Fall back to 'wb' only if `mode is None`

ppolewicz commented on b2sdk/transfer/inbound/downloaded_file.py:
```diff
@@ -157,14 +181,19 @@ def _validate_download(self, bytes_read, actual_sha1):
                 raise TruncatedOutput(bytes_read, desired_length)
 
-    def save(self, file, allow_seeking=True):
+    def save(self, file, allow_seeking: bool | None = None):
```
missing typing for `file`

mjurbanski-reef commented:
replaced by #315
